### DISCUSSION TITLE
Machine Learning Models in CESM Reanalyses

### DISCUSSION THREAD

**Discussion by kdraeder** [NCAR/DART#1043] [Ideas]
This discussion will start as a record of my thoughts about how ML models might be used in the CESM DA context.  My path or strategy so far has been to use the CAMulator as a focus and motivation for learning about ML models, because it appears to be a potentially useful substitute for CAM.  I'm far from an expert in any part of this, or the big picture, except about what was necessary and useful for the CAM6+DART Reanalysis.

This ignores the possibility that the next reanalysis may use a B compset and "strong coupling" DA; active atm, lnd, ocn, sea ice, and probably river components.  I have little idea about the feasibility of using ML models for the non-atm components.  That discussion can happen here, or a related discussion, or in-person.

### Benefits:

- Sufficiently faithful, for DA purposes, to the version of CAM used for training.
- The CAMulator can be trained to predict any (sub)set of prognostic or diagnostic variables on the native model grid.  This includes the "forcing" variables which were the main product of the CAM6+DART Reanalysis. 
- Much cheaper to run (calculations + IO + overhead is 500x cheaper than just CAM's calculations).  See below for details.
- Procedure for developing the emulator is probably straightforward using MILES:CREDIT facility.
- Makes use of GPUs, which are becoming ubiquitous but are still underused.
- It may fit easily into the current interface to DART, using files; DART sees CAM as a black box anyway.

### Drawbacks:

- Must train a new model for each version of the model state. See below for characteristics.
- A sufficient training data set must be found or created.  See Training Tidbits below.
- Training could take days on some fairly small number of nodes, so it may not be cost effective for short assimilations 
- Sufficient and convenient verification metrics would need to be developed.

We would want to train a CAMulator differently from the original.  The focus there was on reproducing multi-year behavior - climatology and large scale slow patterns such as ENSO - while we would want to optimize short term fidelity; hours.  Apparently this training would put less emphasis on multistep training in order to maximize sharpness and variability.

### Value of a faster CAM:
This becomes more valuable for the CESM3 "workhorse" configuration; 

- Spectral Element dynamical core with a nominal 1 deg resolution for atmosphere and land (ne30) 
- 93 levels below the ~80 km model top ("mid-top", resolving the stratosphere and lower mesophere).  We might be able to use the low-top 58 levels below 40 km, if the QBO is not important.  The obs might enforce a QBO anyway.  Even this is 2x the cost of CAM6-FV (32 levels) which doesn't account for other upgrades in CAM7.
- More prognostic chemistry?

### Model states are functions of:

-   prognostic fields; dynamics, chemistry, hydro, ...
-   grid resolution; horizontal and vertical,
-   assimilation frequency; 6 hours has been standard, but the higher tops of CESM3 models might require more frequent, based on WACCM(-X) experience.
-   external forcing of CAM: SST + sea ice extent, green house gases, aerosols, ...

Inputs to the CAMulator include

-   prognostic (T), and
-   dynamic and static (surface geop. h.) forcing fields.

Outputs include the same prognostic fields but no forcing fields, plus > a dozen diagnostic fields. I think DART's perspective will be that;

-   prognostic variables are those which are input and output.
-   diagnostics are only output.
-   the input-only fields are just model forcings.

### Training tidbits:
Related studies have used ~35 years of a single CAM run.  
An ensemble with a shorter, more focused time span might be fine, but would include fewer of the slow process states and make out-of-sample more frequent. 
It depends on how "out of sample" the inputs might be and can be tolerated.

Channel weighting: Recent work has found it beneficial to carefully weight channels (different weather variables at different vertical levels) in the loss function during training (Nguyen et al. 2023). In particular, the method first pioneered by Lam et al. (2023) of down-weighting as pressure level decreases (higher vertical levels are weighted less) and down-weighting according to the standard deviation of temporal differences (𝜎𝛿X) has been found to work well.

### References:

CAMulater Version 1: Fully Resolved, Fast Emulation of the Community Atmosphere Model
William E. Chapman, John S. Schreck, Yingkai Sha, David John Gagne II, Dhamma Kimpara, Laure Zanna, Judith Berner
  
Community Research Earth Digital Intelligence Twin (CREDIT). 
John S. Schreck, Yingkai Sha, William Chapman, Dhamma Kimpara,
Judith Berner, Seth McGinnis, Arnold Kazadi, Negin Sobhani, Ben Kirk, David John Gagne II

Analyzing and Exploring Training Recipes for Large-scale Transformer-based Weather Prediction.
Jared D. Willard, Peter Harrington, Shashank Subramanian, Ankur Mahesh, Travis A. O'Brien,
William D. Collins

https://developers.google.com/machine-learning/glossary

**Comment by kdraeder** [comment]
Here are additional papers and details about the ones listed above.  See also #1091.

- Shreck et al. "COMMUNITY RESEARCH EARTH DIGITAL INTELLIGENCE TWIN (CREDIT) ", arXiv:2411.07814v1 [cs.AI] 9 Nov 2024
- Wang et al. "CROSSFORMER: A VERSATILE VISION TRANSFORMER HINGING ON CROSS-SCALE ATTENTION",2022. URL https://openreview.net/forum?id=_PHymLIxuI
- Vaswani et al. "Attention Is All You Need", arXiv:1706.03762v7 [cs.CL] 2 Aug 2023

Papers that might be helpful to others:

- Kingma and Welling, "Auto-Encoding Variational Bayes", https://arxiv.org/pdf/1312.6114
- Hinton and Salakhutdinov, "Reducing the Dimensionality of Data with Neural Networks" 28 JULY 2006 VOL 313 SCIENCE www.sciencemag.org
- Willard et al. "ANALYZING AND EXPLORING TRAINING RECIPES FOR LARGE-SCALE TRANSFORMER-BASED WEATHER PREDICTION", arXiv:2404.19630v1 [cs.LG] 30 Apr 2024

**Comment by kdraeder** [comment]
Here's the presentation I made in the Projects meeting today.
[project_talk.pptx](https://github.com/user-attachments/files/27257925/project_talk.pptx)

Discussions involved 

- whether CAMulator might make using EnOI unnecessary,
- whether there are cheaper methods than developing an AI model for representing an atmospheric state in a lower dimensional,
- whether there are more effective methods (learned instead of specified) than the SLDA mechanism in Fan et al.,
- whether any kind of physical interpretation of Fan's embedded vectors, which seem to preserve some spatial organization,
- why V seems to be so disconnected from other fields in CREDIT,

Feel free to add topics I missed.
